# DataFrames

[additional learning resources](https://www.w3schools.com/python/pandas/default.asp)

Throughout the course, most of our data exploration will be done with DataFrames. DataFrames are an extremely powerful tool and a natural extension of the Pandas Series. By definition all a DataFrame is:

**A Pandas DataFrame consists of multiple Pandas Series that share index values.**

## Imports

In [2]:
import numpy as np
import pandas as pd

### Creating a DataFrame from Python Objects

In [3]:
# Make sure the seed is in the same cell as the random call
# https://stackoverflow.com/questions/21494489/what-does-numpy-random-seed0-do
np.random.seed(101)
mydata = np.random.randint(0,101,(4,3))

In [4]:
mydata

array([[95, 11, 81],
       [70, 63, 87],
       [75,  9, 77],
       [40,  4, 63]])

In [5]:
myindex = ['CA','NY','AZ','TX']

In [6]:
mycolumns = ['Jan','Feb','Mar']

In [7]:
df = pd.DataFrame(data=mydata)
df

0   1   2
0  95  11  81
1  70  63  87
2  75   9  77
3  40   4  63

In [8]:
df = pd.DataFrame(data=mydata,index=myindex)
df

0   1   2
CA  95  11  81
NY  70  63  87
AZ  75   9  77
TX  40   4  63

In [9]:
df = pd.DataFrame(data=mydata,index=myindex,columns=mycolumns)
df 

Jan  Feb  Mar
CA   95   11   81
NY   70   63   87
AZ   75    9   77
TX   40    4   63

##  Read CSV
Comma Separated Values files are text files that use commas as field delimeters.<br>
Unless you're running the virtual environment included with the course, you may need to install <tt>xlrd</tt> and <tt>openpyxl</tt>.<br>
In your terminal/command prompt run:

    conda install xlrd
    conda install openpyxl

Then restart Jupyter Notebook.
(or use pip install if you aren't using the Anaconda Distribution)

### Understanding File Paths

You have two options when reading a file with pandas:

1. If your .py file or .ipynb notebook is located in the **exact** same folder location as the .csv file you want to read, simply pass in the file name as a string, for example:
    
        df = pd.read_csv('some_file.csv')
        
2. Pass in the entire file path if you are located in a different directory. The file path must be 100% correct in order for this to work. For example:

        df = pd.read_csv("C:\\Users\\myself\\files\\some_file.csv")

#### Print your current directory file path with pwd

In [10]:
pwd

'/Users/joannalin/coding/python/ClassWithAlex'

#### List the files in your current directory with ls

In [11]:
ls

02-Python Crash Course Exercises.ipynb  04-DataFrames.ipynb


In [12]:
df = pd.read_csv('tips.csv')

FileNotFoundError: [Errno 2] No such file or directory: 'tips.csv'

In [ ]:
df

total_bill   tip     sex smoker   day    time  size  price_per_person  \
0         16.99  1.01  Female     No   Sun  Dinner     2              8.49   
1         10.34  1.66    Male     No   Sun  Dinner     3              3.45   
2         21.01  3.50    Male     No   Sun  Dinner     3              7.00   
3         23.68  3.31    Male     No   Sun  Dinner     2             11.84   
4         24.59  3.61  Female     No   Sun  Dinner     4              6.15   
..          ...   ...     ...    ...   ...     ...   ...               ...   
239       29.03  5.92    Male     No   Sat  Dinner     3              9.68   
240       27.18  2.00  Female    Yes   Sat  Dinner     2             13.59   
241       22.67  2.00    Male    Yes   Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No   Sat  Dinner     2              8.91   
243       18.78  3.00  Female     No  Thur  Dinner     2              9.39   

             Payer Name         CC Number Payment ID  
0    Christy Cunningham  3560325168603410    Sun2959  
1        Douglas Tucker  4478071379779230    Sun4608  
2        Travis Walters  6011812112971322    Sun4458  
3      Nathaniel Harris  4676137647685994    Sun5260  
4          Tonya Carter  4832732618637221    Sun2251  
..                  ...               ...        ...  
239       Michael Avila  5296068606052842    Sat2657  
240      Monica Sanders  3506806155565404    Sat1766  
241          Keith Wong  6011891618747196    Sat3880  
242        Dennis Dixon     4375220550950      Sat17  
243     Michelle Hardin  3511451626698139    Thur672  

[244 rows x 11 columns]

# Let's load our dataset into a Notebook!

----
About this DataSet (in case you are interested)

* Description
    * One waiter recorded information about each tip he received over a period of a few months working in one restaurant. He collected several variables:

* Format
    * A data frame with 244 rows and 7 variables

* Details
    * tip in dollars,
    * bill in dollars,
    * sex of the bill payer,
    * whether there were smokers in the party,
    * day of the week,
    * time of day,
    * size of the party.

In all he recorded 244 tips. The data was reported in a collection of case studies for business statistics (Bryant & Smith 1995).

* References
    * Bryant, P. G. and Smith, M (1995) Practical Data Analysis: Case Studies in Business Statistics. Homewood, IL: Richard D. Irwin Publishing:
    
* Note: We created some additional columns with Fake data, including Name, CC Number, and Payment ID.

----

# DataFrames

## Obtaining Basic Information About DataFrame

In [ ]:
df.columns

Index(['total_bill', 'tip', 'sex', 'smoker', 'day', 'time', 'size',
       'price_per_person', 'Payer Name', 'CC Number', 'Payment ID'],
      dtype='object')

In [ ]:
df.index

RangeIndex(start=0, stop=244, step=1)

In [ ]:
df.head(3)

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608  
2      Travis Walters  6011812112971322    Sun4458

In [ ]:
df.tail(3)

total_bill   tip     sex smoker   day    time  size  price_per_person  \
241       22.67  2.00    Male    Yes   Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No   Sat  Dinner     2              8.91   
243       18.78  3.00  Female     No  Thur  Dinner     2              9.39   

          Payer Name         CC Number Payment ID  
241       Keith Wong  6011891618747196    Sat3880  
242     Dennis Dixon     4375220550950      Sat17  
243  Michelle Hardin  3511451626698139    Thur672

In [ ]:
len(df)

244

## Selection and Indexing

Let's learn how to retrieve information from a DataFrame.

### COLUMNS

We will begin be learning how to extract information based on the columns

In [ ]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608  
2      Travis Walters  6011812112971322    Sun4458  
3    Nathaniel Harris  4676137647685994    Sun5260  
4        Tonya Carter  4832732618637221    Sun2251

#### Grab a Single Column

In [ ]:
df['total_bill']

0      16.99
1      10.34
2      21.01
3      23.68
4      24.59
       ...  
239    29.03
240    27.18
241    22.67
242    17.82
243    18.78
Name: total_bill, Length: 244, dtype: float64

In [ ]:
type(df['total_bill'])

pandas.core.series.Series

#### Grab Multiple Columns

In [ ]:
# Note how its a python list of column names! Thus the double brackets.
df[['total_bill','tip']]

total_bill   tip
0         16.99  1.01
1         10.34  1.66
2         21.01  3.50
3         23.68  3.31
4         24.59  3.61
..          ...   ...
239       29.03  5.92
240       27.18  2.00
241       22.67  2.00
242       17.82  1.75
243       18.78  3.00

[244 rows x 2 columns]

#### Create New Columns

In [ ]:
df['tip_percentage'] = 100* df['tip'] / df['total_bill']

In [ ]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  tip_percentage  
0  Christy Cunningham  3560325168603410    Sun2959        5.944673  
1      Douglas Tucker  4478071379779230    Sun4608       16.054159  
2      Travis Walters  6011812112971322    Sun4458       16.658734  
3    Nathaniel Harris  4676137647685994    Sun5260       13.978041  
4        Tonya Carter  4832732618637221    Sun2251       14.680765

In [ ]:
df['price_per_person'] = df['total_bill'] / df['size']

In [ ]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2          8.495000   
1       10.34  1.66    Male     No  Sun  Dinner     3          3.446667   
2       21.01  3.50    Male     No  Sun  Dinner     3          7.003333   
3       23.68  3.31    Male     No  Sun  Dinner     2         11.840000   
4       24.59  3.61  Female     No  Sun  Dinner     4          6.147500   

           Payer Name         CC Number Payment ID  tip_percentage  
0  Christy Cunningham  3560325168603410    Sun2959        5.944673  
1      Douglas Tucker  4478071379779230    Sun4608       16.054159  
2      Travis Walters  6011812112971322    Sun4458       16.658734  
3    Nathaniel Harris  4676137647685994    Sun5260       13.978041  
4        Tonya Carter  4832732618637221    Sun2251       14.680765

#### Adjust Existing Columns

In [ ]:
# Because pandas is based on numpy, we get awesome capabilities with numpy's universal functions!
df['price_per_person'] = np.round(df['price_per_person'],2)

In [ ]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  tip_percentage  
0  Christy Cunningham  3560325168603410    Sun2959        5.944673  
1      Douglas Tucker  4478071379779230    Sun4608       16.054159  
2      Travis Walters  6011812112971322    Sun4458       16.658734  
3    Nathaniel Harris  4676137647685994    Sun5260       13.978041  
4        Tonya Carter  4832732618637221    Sun2251       14.680765

#### Remove Columns

In [ ]:
# df.drop('tip_percentage',axis=1)

In [ ]:
df = df.drop("tip_percentage",axis=1)

In [ ]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608  
2      Travis Walters  6011812112971322    Sun4458  
3    Nathaniel Harris  4676137647685994    Sun5260  
4        Tonya Carter  4832732618637221    Sun2251

# Index Basics

Before going over the same retrieval tasks for rows, let's build some basic understanding of the pandas DataFrame Index.

In [ ]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608  
2      Travis Walters  6011812112971322    Sun4458  
3    Nathaniel Harris  4676137647685994    Sun5260  
4        Tonya Carter  4832732618637221    Sun2251

In [ ]:
df.index

RangeIndex(start=0, stop=244, step=1)

In [ ]:
df.set_index('Payment ID')

total_bill   tip     sex smoker   day    time  size  \
Payment ID                                                        
Sun2959          16.99  1.01  Female     No   Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No   Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No   Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No   Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No   Sun  Dinner     4   
...                ...   ...     ...    ...   ...     ...   ...   
Sat2657          29.03  5.92    Male     No   Sat  Dinner     3   
Sat1766          27.18  2.00  Female    Yes   Sat  Dinner     2   
Sat3880          22.67  2.00    Male    Yes   Sat  Dinner     2   
Sat17            17.82  1.75    Male     No   Sat  Dinner     2   
Thur672          18.78  3.00  Female     No  Thur  Dinner     2   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sun2959                 8.49  Christy Cunningham  3560325168603410  
Sun4608                 3.45      Douglas Tucker  4478071379779230  
Sun4458                 7.00      Travis Walters  6011812112971322  
Sun5260                11.84    Nathaniel Harris  4676137647685994  
Sun2251                 6.15        Tonya Carter  4832732618637221  
...                      ...                 ...               ...  
Sat2657                 9.68       Michael Avila  5296068606052842  
Sat1766                13.59      Monica Sanders  3506806155565404  
Sat3880                11.34          Keith Wong  6011891618747196  
Sat17                   8.91        Dennis Dixon     4375220550950  
Thur672                 9.39     Michelle Hardin  3511451626698139  

[244 rows x 10 columns]

In [ ]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608  
2      Travis Walters  6011812112971322    Sun4458  
3    Nathaniel Harris  4676137647685994    Sun5260  
4        Tonya Carter  4832732618637221    Sun2251

In [ ]:
df = df.set_index('Payment ID')

In [ ]:
df.head()

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No  Sun  Dinner     4   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sun2959                 8.49  Christy Cunningham  3560325168603410  
Sun4608                 3.45      Douglas Tucker  4478071379779230  
Sun4458                 7.00      Travis Walters  6011812112971322  
Sun5260                11.84    Nathaniel Harris  4676137647685994  
Sun2251                 6.15        Tonya Carter  4832732618637221

In [ ]:
df = df.reset_index()

In [ ]:
df.head()

Payment ID  total_bill   tip     sex smoker  day    time  size  \
0    Sun2959       16.99  1.01  Female     No  Sun  Dinner     2   
1    Sun4608       10.34  1.66    Male     No  Sun  Dinner     3   
2    Sun4458       21.01  3.50    Male     No  Sun  Dinner     3   
3    Sun5260       23.68  3.31    Male     No  Sun  Dinner     2   
4    Sun2251       24.59  3.61  Female     No  Sun  Dinner     4   

   price_per_person          Payer Name         CC Number  
0              8.49  Christy Cunningham  3560325168603410  
1              3.45      Douglas Tucker  4478071379779230  
2              7.00      Travis Walters  6011812112971322  
3             11.84    Nathaniel Harris  4676137647685994  
4              6.15        Tonya Carter  4832732618637221

### ROWS

Let's now explore these same concepts but with Rows.

In [ ]:
df.head()

Payment ID  total_bill   tip     sex smoker  day    time  size  \
0    Sun2959       16.99  1.01  Female     No  Sun  Dinner     2   
1    Sun4608       10.34  1.66    Male     No  Sun  Dinner     3   
2    Sun4458       21.01  3.50    Male     No  Sun  Dinner     3   
3    Sun5260       23.68  3.31    Male     No  Sun  Dinner     2   
4    Sun2251       24.59  3.61  Female     No  Sun  Dinner     4   

   price_per_person          Payer Name         CC Number  
0              8.49  Christy Cunningham  3560325168603410  
1              3.45      Douglas Tucker  4478071379779230  
2              7.00      Travis Walters  6011812112971322  
3             11.84    Nathaniel Harris  4676137647685994  
4              6.15        Tonya Carter  4832732618637221

In [ ]:
df = df.set_index('Payment ID')

In [ ]:
df.head()

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No  Sun  Dinner     4   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sun2959                 8.49  Christy Cunningham  3560325168603410  
Sun4608                 3.45      Douglas Tucker  4478071379779230  
Sun4458                 7.00      Travis Walters  6011812112971322  
Sun5260                11.84    Nathaniel Harris  4676137647685994  
Sun2251                 6.15        Tonya Carter  4832732618637221

#### Grab a Single Row

In [ ]:
# Integer Based
df.iloc[0]

total_bill                       16.99
tip                               1.01
sex                             Female
smoker                              No
day                                Sun
time                            Dinner
size                                 2
price_per_person                  8.49
Payer Name          Christy Cunningham
CC Number             3560325168603410
Name: Sun2959, dtype: object

In [ ]:
# Name Based
df.loc['Sun2959']

total_bill                       16.99
tip                               1.01
sex                             Female
smoker                              No
day                                Sun
time                            Dinner
size                                 2
price_per_person                  8.49
Payer Name          Christy Cunningham
CC Number             3560325168603410
Name: Sun2959, dtype: object

#### Grab Multiple Rows

In [ ]:
df.iloc[0:4]

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sun2959                 8.49  Christy Cunningham  3560325168603410  
Sun4608                 3.45      Douglas Tucker  4478071379779230  
Sun4458                 7.00      Travis Walters  6011812112971322  
Sun5260                11.84    Nathaniel Harris  4676137647685994

In [ ]:
df.loc[['Sun2959','Sun5260']]

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sun2959                 8.49  Christy Cunningham  3560325168603410  
Sun5260                11.84    Nathaniel Harris  4676137647685994

#### Remove Row

Typically are datasets will be large enough that we won't remove rows like this since we won't know thier row location for some specific condition, instead, we drop rows based on conditions such as missing data or column values. The next lecture will cover this in a lot more detail.

In [ ]:
df.head()

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No  Sun  Dinner     4   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sun2959                 8.49  Christy Cunningham  3560325168603410  
Sun4608                 3.45      Douglas Tucker  4478071379779230  
Sun4458                 7.00      Travis Walters  6011812112971322  
Sun5260                11.84    Nathaniel Harris  4676137647685994  
Sun2251                 6.15        Tonya Carter  4832732618637221

In [ ]:
df.drop('Sun2959',axis=0).head()

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No  Sun  Dinner     4   
Sun9679          25.29  4.71    Male     No  Sun  Dinner     4   

            price_per_person        Payer Name         CC Number  
Payment ID                                                        
Sun4608                 3.45    Douglas Tucker  4478071379779230  
Sun4458                 7.00    Travis Walters  6011812112971322  
Sun5260                11.84  Nathaniel Harris  4676137647685994  
Sun2251                 6.15      Tonya Carter  4832732618637221  
Sun9679                 6.32        Erik Smith   213140353657882

In [ ]:
# Error if you have a named index!
# df.drop(0,axis=0).head()

--------

# Conditional Filter

## Single Condition

In [ ]:
bool_series = df['tip'] > 5
bool_series

0      False
1      False
2      False
3      False
4      False
       ...  
239     True
240    False
241    False
242    False
243    False
Name: tip, Length: 244, dtype: bool

In [ ]:
df[bool_series]

total_bill    tip     sex smoker   day    time  size  price_per_person  \
23        39.42   7.58    Male     No   Sat  Dinner     4              9.86   
44        30.40   5.60    Male     No   Sun  Dinner     4              7.60   
47        32.40   6.00    Male     No   Sun  Dinner     4              8.10   
52        34.81   5.20  Female     No   Sun  Dinner     4              8.70   
59        48.27   6.73    Male     No   Sat  Dinner     4             12.07   
85        34.83   5.17  Female     No  Thur   Lunch     4              8.71   
88        24.71   5.85    Male     No  Thur   Lunch     2             12.36   
116       29.93   5.07    Male     No   Sun  Dinner     4              7.48   
141       34.30   6.70    Male     No  Thur   Lunch     6              5.72   
155       29.85   5.14  Female     No   Sun  Dinner     5              5.97   
170       50.81  10.00    Male    Yes   Sat  Dinner     3             16.94   
172        7.25   5.15    Male    Yes   Sun  Dinner     2              3.62   
181       23.33   5.65    Male    Yes   Sun  Dinner     2             11.66   
183       23.17   6.50    Male    Yes   Sun  Dinner     4              5.79   
211       25.89   5.16    Male    Yes   Sat  Dinner     4              6.47   
212       48.33   9.00    Male     No   Sat  Dinner     4             12.08   
214       28.17   6.50  Female    Yes   Sat  Dinner     3              9.39   
239       29.03   5.92    Male     No   Sat  Dinner     3              9.68   

            Payer Name         CC Number Payment ID  
23      Lance Peterson  3542584061609808     Sat239  
44         Todd Cooper      503846761263    Sun2274  
47        James Barnes  3552002592874186    Sun9677  
52        Emily Daniel  4291280793094374    Sun6165  
59         Brian Ortiz  6596453823950595    Sat8139  
85         Shawna Cook  6011787464177340   Thur7972  
88        Roger Taylor     4410248629955   Thur9003  
116        Shawn Blake  4689079711213722      Sun22  
141     Steven Carlson  3526515703718508   Thur1025  
155     Madison Wilson  4210875236164664    Sun9176  
170      Gregory Clark  5473850968388236    Sat1954  
172        Larry White    30432617123103    Sun9209  
181          Jason Cox  6556931703586223    Sun3402  
183  Dr. Michael James     4718501859162    Sun6059  
211     Christopher Li  6011962464150569    Sat6735  
212    Alex Williamson      676218815212    Sat4590  
214    Marissa Jackson  4922302538691962    Sat3374  
239      Michael Avila  5296068606052842    Sat2657

In [ ]:
df[df['tip'] > 5]

NameError: name 'df' is not defined

In [ ]:
df[df['sex'] == 'Male']

total_bill   tip   sex smoker  day    time  size  price_per_person  \
1         10.34  1.66  Male     No  Sun  Dinner     3              3.45   
2         21.01  3.50  Male     No  Sun  Dinner     3              7.00   
3         23.68  3.31  Male     No  Sun  Dinner     2             11.84   
5         25.29  4.71  Male     No  Sun  Dinner     4              6.32   
6          8.77  2.00  Male     No  Sun  Dinner     2              4.38   
..          ...   ...   ...    ...  ...     ...   ...               ...   
236       12.60  1.00  Male    Yes  Sat  Dinner     2              6.30   
237       32.83  1.17  Male    Yes  Sat  Dinner     2             16.42   
239       29.03  5.92  Male     No  Sat  Dinner     3              9.68   
241       22.67  2.00  Male    Yes  Sat  Dinner     2             11.34   
242       17.82  1.75  Male     No  Sat  Dinner     2              8.91   

             Payer Name         CC Number Payment ID  
1        Douglas Tucker  4478071379779230    Sun4608  
2        Travis Walters  6011812112971322    Sun4458  
3      Nathaniel Harris  4676137647685994    Sun5260  
5            Erik Smith   213140353657882    Sun9679  
6    Kristopher Johnson  2223727524230344    Sun5985  
..                  ...               ...        ...  
236       Matthew Myers  3543676378973965    Sat5032  
237        Thomas Brown  4284722681265508    Sat2929  
239       Michael Avila  5296068606052842    Sat2657  
241          Keith Wong  6011891618747196    Sat3880  
242        Dennis Dixon     4375220550950      Sat17  

[157 rows x 11 columns]

## Multiple-condition

In [ ]:
df[(df['total_bill'] > 30) & (df['sex']=='Male')]

total_bill    tip   sex smoker   day    time  size  price_per_person  \
23        39.42   7.58  Male     No   Sat  Dinner     4              9.86   
39        31.27   5.00  Male     No   Sat  Dinner     3             10.42   
44        30.40   5.60  Male     No   Sun  Dinner     4              7.60   
47        32.40   6.00  Male     No   Sun  Dinner     4              8.10   
56        38.01   3.00  Male    Yes   Sat  Dinner     4              9.50   
59        48.27   6.73  Male     No   Sat  Dinner     4             12.07   
83        32.68   5.00  Male    Yes  Thur   Lunch     2             16.34   
95        40.17   4.73  Male    Yes   Fri  Dinner     4             10.04   
112       38.07   4.00  Male     No   Sun  Dinner     3             12.69   
141       34.30   6.70  Male     No  Thur   Lunch     6              5.72   
142       41.19   5.00  Male     No  Thur   Lunch     5              8.24   
156       48.17   5.00  Male     No   Sun  Dinner     6              8.03   
167       31.71   4.50  Male     No   Sun  Dinner     4              7.93   
170       50.81  10.00  Male    Yes   Sat  Dinner     3             16.94   
173       31.85   3.18  Male    Yes   Sun  Dinner     2             15.92   
175       32.90   3.11  Male    Yes   Sun  Dinner     2             16.45   
179       34.63   3.55  Male    Yes   Sun  Dinner     2             17.32   
180       34.65   3.68  Male    Yes   Sun  Dinner     4              8.66   
182       45.35   3.50  Male    Yes   Sun  Dinner     3             15.12   
184       40.55   3.00  Male    Yes   Sun  Dinner     2             20.27   
187       30.46   2.00  Male    Yes   Sun  Dinner     5              6.09   
207       38.73   3.00  Male    Yes   Sat  Dinner     4              9.68   
210       30.06   2.00  Male    Yes   Sat  Dinner     3             10.02   
212       48.33   9.00  Male     No   Sat  Dinner     4             12.08   
237       32.83   1.17  Male    Yes   Sat  Dinner     2             16.42   

                Payer Name         CC Number Payment ID  
23          Lance Peterson  3542584061609808     Sat239  
39       Mr. Brandon Berry  6011525851069856    Sat6373  
44             Todd Cooper      503846761263    Sun2274  
47            James Barnes  3552002592874186    Sun9677  
56   James Christensen DDS   349793629453226    Sat8903  
59             Brian Ortiz  6596453823950595    Sat8139  
83           Daniel Murphy  5356177501009133   Thur8801  
95           Aaron Bentley   180026611638690    Fri9628  
112             Jeff Lopez  3572865915176463     Sun591  
141         Steven Carlson  3526515703718508   Thur1025  
142           Eric Andrews  4356531761046453   Thur3621  
156          Ryan Gonzales  3523151482063321    Sun7518  
167         Michael Lawson  3566285921227119    Sun3719  
170          Gregory Clark  5473850968388236    Sat1954  
173            Scott Perez  3577115550328507    Sun9335  
175        Nathan Reynolds   370307040837149    Sun5109  
179           Brian Bailey   346656312114848    Sun9851  
180       James Hebert DDS      676168737648    Sun7544  
182           Jose Parsons  4112207559459910    Sun2337  
184            Stephen Cox  3547798222044029    Sun5140  
187          David Barrett  4792882899700988    Sun9987  
207          Ricky Ramirez   347817964484033    Sat4505  
210          Shawn Mendoza    30184049218122    Sat8361  
212        Alex Williamson      676218815212    Sat4590  
237           Thomas Brown  4284722681265508    Sat2929

In [ ]:
# The Weekend
df[(df['day'] =='Sun') | (df['day']=='Sat')]

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0         16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1         10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2         21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3         23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4         24.59  3.61  Female     No  Sun  Dinner     4              6.15   
..          ...   ...     ...    ...  ...     ...   ...               ...   
238       35.83  4.67  Female     No  Sat  Dinner     3             11.94   
239       29.03  5.92    Male     No  Sat  Dinner     3              9.68   
240       27.18  2.00  Female    Yes  Sat  Dinner     2             13.59   
241       22.67  2.00    Male    Yes  Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No  Sat  Dinner     2              8.91   

             Payer Name         CC Number Payment ID  
0    Christy Cunningham  3560325168603410    Sun2959  
1        Douglas Tucker  4478071379779230    Sun4608  
2        Travis Walters  6011812112971322    Sun4458  
3      Nathaniel Harris  4676137647685994    Sun5260  
4          Tonya Carter  4832732618637221    Sun2251  
..                  ...               ...        ...  
238      Kimberly Crane      676184013727    Sat9777  
239       Michael Avila  5296068606052842    Sat2657  
240      Monica Sanders  3506806155565404    Sat1766  
241          Keith Wong  6011891618747196    Sat3880  
242        Dennis Dixon     4375220550950      Sat17  

[163 rows x 11 columns]